staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go

```diff
@@ -163,7 +163,40 @@ func (g *Cloud) InstanceShutdownByProviderID(ctx context.Context, providerID str
 
 // InstanceMetadataByProviderID returns metadata of the specified instance.
 func (g *Cloud) InstanceMetadataByProviderID(ctx context.Context, providerID string) (*cloudprovider.InstanceMetadata, error) {
-	return nil, fmt.Errorf("unimplemented")
+	md := &cloudprovider.InstanceMetadata{}
+	var err error
+
+	if providerID == "" {
+		return nil, fmt.Errorf("no providerID")
+	}
+
+	md.ProviderID = providerID
+	timeoutCtx, cancel := context.WithTimeout(ctx, 1*time.Hour)
+	defer cancel()
+
+	_, zone, name, err := splitProviderID(providerID)
+	if err != nil {
+		return nil, err
+	}
+
+	instance, err := g.c.Instances().Get(timeoutCtx, meta.ZonalKey(canonicalizeInstanceName(name), zone))
+	if err != nil {
+		return nil, fmt.Errorf("error while querying for providerID %q: %v", providerID, err)
+	}
+
+	if len(instance.NetworkInterfaces) < 1 {
```

Comment:
Can you please factor this out to a separate function (nodeAddressesFromGCEInstance) and share it between this and the NodeAddressesByProviderID functions?

Reply:
It seems the same comment holds for some other cloud providers (vsphere is what I looked at).